### ISSUE TITLE
Per-class strategy

### ISSUE THREAD

**Issue by pmelchior** [joshua-lintropic/PFS-Class-Optimizer#1]
The current GE survey relies on a cost-based optimizer, but what's fundamentally desired is a more complex optimization that balances the completeness of multiple galaxy classes. It is therefore sensible to first establish the best numbers across all classes that could be achieved under ideal conditions. In short: the optimal strategy we will then aim to follow.

That can be expressed as solving

```math
\begin{align}
&\mathrm{argmax}_{(n_{11}, \dots, n_{IK})} \min\left(\frac{n_1}{N_1}, \dots,\frac{n_{12}}{N_{12}}\right)\\
&n_{i} \equiv \sum_k n_{ik}\\
&\forall k: \sum_i n_{ik}T_i - T \leq 0\\
&\forall i,k: n_{ik} \in \mathbb{N}_0^+
\end{align}
```

This problem can be optimized with GD by minimizing the loss function

```math
\begin{align}
L(n'_{11}, \dots, n'_{IK}) &=-\min\left(\frac{n_1}{N_1}, \dots,\frac{n_{12}}{N_{12}}\right) + \lambda \sum_k p (\sum_i n_{ik}T_i - T)\\
\text{with}\ n_{ik} &= \text{round}(n'_{ik})
\end{align}
```

where the round function is replaced during training by the noisy sigmoid function (see eq. 5 in Tianshu's paper).

This approach defines how many sources of class $i$ any fiber $k$ should observe, but not the time-ordering. We could simply order them by required time $T\_i$, so that the long-run targets are done first. This leaves the actual assignment of fiber on galaxy left to step 2.

There’s one foreseeable problem: degeneracies. The fibers are indistinguishable. So any permutation of $k$ will be equivalent. We will probably need to break these degeneracies to allow the optimizer to find solutions more easily.

There's another problem: We implicitly assume that a fiber can always get the ideal number of targets from each class. We assume infinite supply. That cannot be assured, so our strategy represents the optimal case in which one doesn't run out of sources.
On the position side, source restrictions may help us break the degeneracies above. Once we know the actual target catalogs for the GE fields, we can add the number of targets available in every class as a vector $N_{ik}$ of length $I$ to the properties of the fiber nodes and add another penalty: $\forall i,k: n_{ik} \leq N_{ik}$.

This will ensure that only fibers with sufficient numbers of suitable targets will "claim" them and thus partially break the fiber degeneracies. It may not be enough, though.
